bread/tests/test_public_visible.py

```
@@ -0,0 +1,99 @@
+        url_intpk = {x for x in self.ALL_URL_NAMES if "<int:pk>" in x}
+
+        # update the exceptions for some pages that can be visible to the public
+        # and those pages with dynamic url regex
+        # and those out of the bread's scope (e.g., basxconnect)
+        url_names = self.ALL_URL_NAMES - excepted_urlpatterns - url_intpk
+
+        # manually add a list of urls with a dynamic subdirectory.
+        # (e.g., <int:pk>)
+        for url in url_intpk:
+            url_names.add(url.replace("<int:pk>", str(secrets.randbelow(100))))
+
+        # try fetching each url and see the response
+        c = Client()
+
+        for url in url_names:
+            try:
+                response = c.get(url)
+                # if the status code is 4,
+                # there's some errors loading that pages
+                self.assertTrue(
+                    response.status_code // 100 != 4,
+                    "This page %s failed to load, with status code %d."
+                    % (url, response.status_code),
+                )
+                # if the status code is 3, and the Location is /bread/accounts/login/
+                # it's likely invisible to anonymous users, as the page
+                # redirects to the login page.
+                self.assertTrue(
+                    response.status_code // 100 == 3
+                    and response.headers["Location"].startswith(
+                        "/bread/accounts/login/"
+                    ),
+                    "This page %s may be visible to anonymous users, with the status code %d"
+                    % (url, response.status_code),
+                )
+
+            except TemplateDoesNotExist as e:
+                if hasattr(e, "message"):
+                    print("The template for", url, "at", e.message, "does not exist.")
+                else:
+                    print("The template for", url, "does not exist.")
+
+                traceback.print_exc()
```

Hmmm... Where is this exception coming from? Can you explain why we need to catch it?

---

There's one exception raised while testing.

```
django.template.exceptions.TemplateDoesNotExist: registration/login.html
```

---

Hmm... How many of these do we have?

If it is only from login or auth views we can catch the exception and silently drop them.

---

Only one here for now.

---

Okay, then I suggest we add an exception for that specific URL. But other URLs should still raise the TemplateDoesNotExist exception.